File: radiosim/channel/lib/node_proxy.cpp
#include <iostream>
#include <string>
#include <vector>
#include <thread>
#include <complex>
#include <chrono>
#include <zmq.hpp>
#include <node_proxy.h>


// Constructor
NodeProxy::NodeProxy(zmq::context_t& ctx, unsigned int txport, unsigned int rxport, int buffer_size){
	init_sockets(ctx, txport, rxport);
	init_buffers(buffer_size);
}


// Initialize sockets
void NodeProxy::init_sockets(zmq::context_t& ctx, unsigned int txport, unsigned int rxport){
	this->txport = txport;
	this->rxport = rxport;

	txsocket = zmq::socket_t(ctx, zmq::socket_type::pull);
	txsocket.bind( "tcp://127.0.0.1:" + std::to_string(txport) );

	rxsocket = zmq::socket_t(ctx, zmq::socket_type::push);
	rxsocket.bind( "tcp://127.0.0.1:" + std::to_string(rxport) );
}


// Initialize buffers
void NodeProxy::init_buffers(int buffer_size){
	//txbuffer.reserve(buffer_size);
	//rxbuffer.reserve(buffer_size);
}


// Startup node proxy threads
void NodeProxy::start(){
	std::thread txlisten_thread( &NodeProxy::txlisten, this);
	std::thread rxsend_thread( &NodeProxy::rxsend, this);
	txlisten_thread.detach();
	rxsend_thread.detach();
}


// Listen on txsocket for data
void NodeProxy::txlisten(){
	std::cout << "NodeProxy listening on localhost:" << txport << std::endl;

	while (true) {
		zmq::message_t msg;
		txsocket.recv( msg, zmq::recv_flags::none );
		vector_c64 data = unpack_to_complex64(msg);
		txbuffer.push( data );
	}
}


// Unpack message bytes to vector of complex64 data
inline vector_c64 NodeProxy::unpack_to_complex64(zmq::message_t& msg){
	size_t nbytes = msg.size();
	vector_c64 complexdata( nbytes/2/sizeof(float) );
	memcpy(complexdata.data(), msg.data(), nbytes);
	return complexdata;
}


// Send channel data out rxsocket
void NodeProxy::rxsend(){
	std::cout << "NodeProxy sending from localhost:" << rxport << std::endl;
	vector_c64 data;
	data.reserve(512);
	while (true){
		std::this_thread::sleep_for(std::chrono::milliseconds(100));
		if (not rxbuffer.empty()){
			data = rxbuffer.front();
			rxbuffer.pop();
			zmq::message_t msg = pack_complex64_to_message(data);
			rxsocket.send( msg, zmq::send_flags::none );
		}
	}
}



// Write ZMQ message object from vector of complex64 data
inline zmq::message_t NodeProxy::pack_complex64_to_message(vector_c64& complexdata){
	size_t nbytes = 2*complexdata.size()*sizeof(float);
	zmq::message_t msg(nbytes);
	memcpy(msg.data(), complexdata.data(), nbytes);
	return msg;
}

File: radiosim/channel/python/node_proxy.py
import numpy as np
import threading
import zmq
import struct
import queue
import logging
log = logging.getLogger(__name__)

 
class NodeProxy:
	def __init__(self, channel, txport=None, rxport=None, buffer_size=1024, timeout=1000):
		self.channel = channel
		self.initialize_buffers( buffer_size )
		self.initialize_sockets( txport, rxport )
		self.timeout = timeout


	def __repr__(self):
		return f"NodeProxy(channel={self.channel.__repr__()}, txport={self.txport}, rxport={self.rxport}, buffer_size={self.buffer_size}, timeout={self.timeout})"


	def initialize_buffers(self, buffer_size):
		self.buffer_size = buffer_size
		self.txbuffer = queue.Queue( maxsize=buffer_size )
		self.rxbuffer = queue.Queue( maxsize=buffer_size )


	def initialize_sockets(self, txport, rxport):
		self.txsocket = None
		self.rxsocket = None

		tcp_lo = "tcp://127.0.0.1"
		if txport is not None:
			self.txsocket = self.channel.ctx.socket( zmq.PULL )
			self.txsocket.connect( f"{tcp_lo}:{txport}" )
		if rxport is not None:
			self.rxsocket = self.channel.ctx.socket( zmq.PUSH )
			self.rxsocket.bind( f"{tcp_lo}:{rxport}" )


	def start(self):
		log.debug("Starting NodeProxy sockets")
		if self.txsocket is not None:
			self._listen_thread = threading.Thread( target=self._listen )
			self._listen_thread.daemon = True
			self._listen_thread.start()
		if self.rxsocket is not None:
			self._send_thread = threading.Thread( target=self._send )
			self._send_thread.daemon = True
			self._send_thread.start()


	def _listen(self):
		log.debug("Begin executing NodeProxy listen thread")
		while True:
			if self.txsocket.poll( self.timeout, zmq.POLLIN ):
				data = self.txsocket.recv_serialized( self._deserialize )
				self.txbuffer.put( data )


	def _deserialize(self, zmq_frames):
		frame = zmq_frames[0]
		num_c64 = len(frame) // 8
		flatdata = np.array(struct.unpack("ff"*num_c64, frame), dtype=np.complex64)
		complex_data = flatdata[::2] + 1j*flatdata[1::2]
		return complex_data


	def _send(self):
		log.debug("Begin executing NodeProxy send thread")
		while True:
			try:
				data = self.rxbuffer.get( self.timeout/1e3 )
			except queue.Empty:
				continue
			self.rxsocket.send( data )
			self.rxbuffer.task_done()
